### PR TITLE
java-profile flamegraphs saved in tmp folder, duplicate jvm arg fix

### PR DESCRIPTION
**Pull request by lancelui-amzn**
*Description of changes:*
- Aperf record creates a folder in /tmp/ to allow JVMs with limited permissions to save flamegraph files. Results are copied to the specified record folder by aperf. Previously, JVMs would directly save folders to the record folder.
- Made record and report integration tests sequential to avoid conflicts in tmp folder
- Changed `CollectorParams` path fields to `PathBuf` from `String`
- Fixed bug in argument parsing that attempted to profile the same JVM twice if duplicate arguments were given.

Testing:

- Profiled groovy and Cassandra benchmarks, both with various permissions.
- Ran updated test suite and validated results by created files and visually through UI

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by wash-amzn** [comment]
> Previously, JVMs would directly save folders to the record folder.

Why was that a problem?

File: src/data/java_profile.rs [hunk 1]
```diff
@@ -211,6 +212,19 @@ impl CollectData for JavaProfileRaw {
...
+            let mut html_path = data_dir.clone();
+            html_path.push(format!("{}-java-flamegraph-{}.html", params.run_name, key));
+
+            let html_loc = html_path.to_str().unwrap();
+            let tmp_loc = format!(
+                "/tmp/aperf_tmp/{}-java-flamegraph-{}.html",
+                params.run_name, key
+            );
+
+            fs::copy(tmp_loc.clone(), html_loc).ok();
```
Comment (wash-amzn) [inline src/data/java_profile.rs:225]: A move would be better than a copy. Although, the answer to my question of why we have to write to /tmp/ in the first place might mean it has to be a copy.

File: src/record.rs [hunk 2]
```diff
@@ -2,6 +2,10 @@ use crate::{data, InitParams, PERFORMANCE_DATA};
 use anyhow::Result;
 use clap::Args;
 use log::{debug, error, info};
+use std::fs;
+use std::os::unix::fs::PermissionsExt;
+
+pub static APERF_TMP: &str = "/tmp/aperf_tmp";
```
Comment (janaknat) [inline src/record.rs:8]: This should be passed in as a parameter.

File: src/data/java_profile.rs [hunk 3]
```diff
@@ -211,6 +212,19 @@ impl CollectData for JavaProfileRaw {
         }
 
         let data_dir = PathBuf::from(params.data_dir.clone());
+        for key in self.process_map.keys() {
+            let mut html_path = data_dir.clone();
+            html_path.push(format!("{}-java-flamegraph-{}.html", params.run_name, key));
+
+            let html_loc = html_path.to_str().unwrap();
+            let tmp_loc = format!(
+                "/tmp/aperf_tmp/{}-java-flamegraph-{}.html",
```
Comment (janaknat) [inline src/data/java_profile.rs:221]: We are hard-coding /tmp/aperf_tmp in multiple locations. We should have the tmp location passed in as a parameter.

**Comment by lancelui-amzn** [comment]
> > Previously, JVMs would directly save folders to the record folder.
> 
> Why was that a problem?

When aperf is launched with sudo, the record folder it creates has sudo permissions. A non-sudo JVM being profiled will not be able to save the flamegraph to this folder.

File: src/record.rs (hunk 2 above)
Comment (lancelui-amzn) [reply src/record.rs:8]: Do you mean like another CLI option and part of collector params?

Comment (janaknat) [reply src/record.rs:8]: As part of the collector params passed into each of the data type.

File: src/data.rs [hunk 4]
```diff
@@ -58,6 +58,7 @@ pub struct CollectorParams {
     pub data_dir: String,
     pub run_name: String,
     pub profile: HashMap<String, String>,
+    pub tmp_dir: String,
```
Comment (janaknat) [inline src/data.rs:61]: Let's use PathBuf for paths. We should change the other members of the CollectorParams from String to PathBuf.

**Review by wash-amzn — APPROVED** [review]
Approved contingent on Janak's approval

**Review by janaknat — APPROVED** [review]

**Pull request merged**